### PR TITLE
Facts: Use vm_stat instead of sysctl for free memory

### LINKED ISSUE #52896: ansible_memfree_mb reported as negative numbes on macos
##### SUMMARY
On MacOS it seems that ansible fails to correctly retrieve the amount of free memory, even if the system has more than enough available.

It seems that the issue can be reproduced with any Python interpreter: I used sytem python 2.7 and also the python 3.6 installed via pyenv and with both I go the same results.

On other platforms I got correct results.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
gather

##### ANSIBLE VERSION

```paste below
ansible 2.7.7
  config file = /Users/ssbarnea/.ansible.cfg
  configured module search path = ['/Users/ssbarnea/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/ssbarnea/.pyenv/versions/3.6.7/lib/python3.6/site-packages/ansible
  executable location = /Users/ssbarnea/.pyenv/versions/3.6.7/bin/ansible
  python version = 3.6.7 (default, Nov 12 2018, 16:11:56) [GCC 4.2.1 Compatible Apple LLVM 10.0.0 (clang-1000.11.45.5)]
```

##### CONFIGURATION

```paste below
N/A
```

##### OS / ENVIRONMENT

MacOS Mojave

##### STEPS TO REPRODUCE

```bash
ansible -v localhost -m setup |grep mem
```

##### EXPECTED RESULTS
A numeric value that is not negative.

##### ACTUAL RESULTS
```
$ ansible -v localhost -m setup |grep mem                                                                                                            "ansible_memfree_mb": -1753,
        "ansible_memtotal_mb": 40960,
```

For convenience I also include a sample output of `vm_stat` command which is the standard way to aquire information about available memory on macos:
```
$ vm_stat                                                                                                                
Mach Virtual Memory Statistics: (page size of 4096 bytes)
Pages free:                              412812.
Pages active:                           4049932.
Pages inactive:                         3996190.
Pages speculative:                        87484.
Pages throttled:                              0.
Pages wired down:                       1001507.
Pages purgeable:                         287560.
"Translation faults":                6349300788.
Pages copy-on-write:                  408554502.
Pages zero filled:                   3057101680.
Pages reactivated:                     48520624.
Pages purged:                          26670474.
File-backed pages:                      2107078.
Anonymous pages:                        6026528.
Pages stored in compressor:             4881797.
Pages occupied by compressor:            937316.
Decompressions:                        38012276.
Compressions:                          65573771.
Pageins:                               55372367.
Pageouts:                                340640.
Swapins:                               22331600.
Swapouts:                              25261950.
```

### PR DESCRIPTION
**Pull request by ssbarnea**
##### SUMMARY
Fixes #52896 
Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/module_utils/facts/hardware/darwin.py

### CODE REVIEW

**Comment by Akasurde** [comment]
cc @samdoran Can we merge this if you are OK with the changes ? Thanks.

**Comment by ssbarnea** [comment]
ready_for_review
@samdoran @bcoca can you help?

**Comment by ssbarnea** [comment]
@Akasurde I have being told that we also need a changelog fragment for this in order to be merged and we need to move quickly to get into Thursday release. 

I think is likely to be looking like:
```
bugfixes:
  - setup - fixed free memory detection on MacoS
```

More info at https://docs.ansible.com/ansible/latest/reference_appendices/release_and_maintenance.html#changelogs-how-to

I would have tryied to add the fragment file myself but I have no push rights to your branch, the one used to create the PR.

File: lib/ansible/module_utils/facts/hardware/darwin.py
Comment (samdoran) [inline lib/ansible/module_utils/facts/hardware/darwin.py]: Use `get_bin_path()` to ensure we always can find the command in case the path changes.
```suggestion
        vm_stat_command = get_bin_path('vm_stat', True)
        rc, out, err = self.module.run_command(vm_stat_command)
```

**Review by samdoran — CHANGES_REQUESTED** [review]
Rather than use a regexp for parsing the command output, we can make the output into a dict and get the keys we want:

```patch
diff --git a/lib/ansible/module_utils/facts/hardware/darwin.py b/lib/ansible/module_utils/facts/hardware/darwin.py
index 065a4d5cf7..280d640410 100644
--- a/lib/ansible/module_utils/facts/hardware/darwin.py
+++ b/lib/ansible/module_utils/facts/hardware/darwin.py
@@ -18,7 +18,7 @@ from __future__ import (absolute_import, division, print_function)
 __metaclass__ = type
 
 import re
-
+from ansible.module_utils.common.process import get_bin_path
 from ansible.module_utils.facts.hardware.base import Hardware, HardwareCollector
 from ansible.module_utils.facts.sysctl import get_sysctl
 
@@ -90,18 +90,30 @@ class DarwinHardware(Hardware):
         }
 
         total_used = 0
-        rc, out, err = self.module.run_command("/usr/bin/vm_stat")
+        vm_stat_command = get_bin_path('vm_stat', True)
+        rc, out, err = self.module.run_command(vm_stat_command)
         if rc == 0:
-            # Free = Total - (Wired + active + inactive)
-            wired_mem_mb = re.search(r'Pages wired down:\s+(\d+)\.', out)
-            active_mem_mb = re.search(r'Pages active:\s+(\d+)\.', out)
-            inactive_mem_mb = re.search(r'Pages inactive:\s+(\d+)\.', out)
-            if wired_mem_mb:
-                total_used += int(wired_mem_mb.group(1)) * 4096
-            if active_mem_mb:
-                total_used += int(active_mem_mb.group(1)) * 4096
-            if inactive_mem_mb:
-                total_used += int(inactive_mem_mb.group(1)) * 4096
+            # Get a generator of tuples from the command output so we can later
+            # turn it into a dictionary
+            memory_stats = (line.rstrip('.').split(':', 1) for line in out.splitlines())
+
+            # Strip extra left spaces from the value
+            memory_stats = dict((k, v.lstrip()) for k, v in memory_stats)
+
+            for k, v in memory_stats.items():
+                try:
+                    memory_stats[k] = int(v)
+                except ValueError as ve:
+                    # Most values convert cleanly to integors but if the field does
+                    # not convert to an integer, just leave it alone.
+                    pass
+
+            if memory_stats.get('Pages wired down'):
+                total_used += memory_stats['Pages wired down'] * 4096
+            if memory_stats.get('Pages active'):
+                total_used += memory_stats['Pages active'] * 4096
+            if memory_stats.get('Pages inactive'):
+                total_used += memory_stats['Pages inactive'] * 4096
 
         memory_facts['memfree_mb'] = memory_facts['memtotal_mb'] - (total_used // 1024 // 1024)
 
```

**Review by samdoran — APPROVED** [review]

File: lib/ansible/module_utils/facts/hardware/darwin.py
Comment (samdoran) [inline lib/ansible/module_utils/facts/hardware/darwin.py]:
```suggestion
                    # Most values convert cleanly to integers but if the field does
```

**Comment by Akasurde** [comment]
@ssbarnea @samdoran Thanks for review.

**Pull request merged**
